### ISSUE TITLE
When Caching is enabled, flushing (e.g. on Dispose) sometimes does not flush all events on app crash

### ISSUE THREAD

**Issue by v1adis1ave** [getsentry/sentry-dotnet#2025]
### Package

Sentry

### .NET Flavor

.NET

### .NET Version

6.0.1

### OS

Any (not platform specific)

### SDK Version

 3.22.0

### Steps to Reproduce

1. Initialize sentry SDK and set `CacheDirectoryPath`
2. Capture event
3. Invoke sentrySdk's `Dispose` from other thread and wait it to complete
4. Exit app

Attached source code using which it's reproducible on MacOS ( I am using Macbook Air 2020 M1 with OS BigSur 11.5.1 )

```csharp
public static class Program
{

    static void Main()
    {
        var sentryDisposableClient = SentrySdk.Init(sentryOptions =>
        {
            sentryOptions.Dsn = // path to dsn
            sentryOptions.Release = "0.0.0.1";
            sentryOptions.Environment = "Debug";
            sentryOptions.CacheDirectoryPath = AppDomain.CurrentDomain.BaseDirectory;
            sentryOptions.IsGlobalModeEnabled = true;
            sentryOptions.ShutdownTimeout = TimeSpan.FromSeconds(10);
            sentryOptions.InitCacheFlushTimeout = TimeSpan.Zero;
            sentryOptions.StackTraceMode = StackTraceMode.Original;
            sentryOptions.DisableAppDomainProcessExitFlush();
            sentryOptions.DisableAppDomainUnhandledExceptionCapture();
            sentryOptions.AutoSessionTracking = true;
        });

        var capturedEventId = SentrySdk.CaptureEvent(
            new SentryEvent(new System.Exception("description to easily find this particular exception"))
            {
                Level = SentryLevel.Fatal
            });

        InvokeAsBackgroundTask(sentryDisposableClient.Dispose)
            .Wait(TimeSpan.FromSeconds(10));

        Console.WriteLine($"Failing fast after capturing {capturedEventId}");

        Environment.FailFast("1");
    }
    
    static Task InvokeAsBackgroundTask(Action action)
    {
        var taskCompletionSource = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);

        var backgroundThread = new Thread(() =>
        {
            try
            {
                action.Invoke();
                taskCompletionSource.TrySetResult();
            }
            catch (Exception ex)
            {
                taskCompletionSource.TrySetException(ex);
            }
        });
        backgroundThread.IsBackground = true;
        backgroundThread.Start();

        return taskCompletionSource.Task;
    }
}

```

### Expected Result

Logs are sent on application crash ( just before the app finishes execution )

### Actual Result

Logs are NOT sent on application crash, and `.envelope` file is created in `%CACHE_DIR%/%dsn.hash%/_processing` soon after app is exited.

**Comment by mattjohnsonpint** [comment]
Hi.  Thanks for reporting.  

I tried your sample code without modification using the version of the SDK you stated and it worked on the first try.

<img width="445" alt="image" src="https://user-images.githubusercontent.com/1396388/198740662-dc9128e4-2130-4809-a118-a8c1cfecec64.png">

Just to double check - are you sure the version is 3.22.0 and not an earlier version?  We fixed a related issue in 3.21.0 (see #1883 / #1890).

Assuming indeed 3.22.0, then I will keep investigating.  Please also tell me, are you running this from an IDE, or on the command line with `dotnet run`, or some other method? And does it matter if you use a debug or release configuration for your app?  Thanks.

**Comment by mattjohnsonpint** [comment]
Actually, I was indeed able to reproduce - just with some intermittency.  I'll get to this ASAP.  Thanks!

**Comment by mattjohnsonpint** [comment]
I have identified the cause and have a fix.  See #2026.  It will be released soon.

Thanks.

**Comment by mattjohnsonpint** [comment]
@v1adis1ave - This issue should be resolved with release 3.23.1.  Please update and confirm.  Thanks.

**Comment by v1adis1ave** [comment]
@mattjohnsonpint Tried with version 3.23.1, issue not reproduced.

Many thanks for prompt reply and addressing the issue.
